ooni/tests/test_scheduler.py

```diff
@@ -51,8 +52,40 @@ def task(subself):
             self.assertEqual(len(in_file.readlines()), 1)
 
         self.assertEqual(dummy_st.should_run, False)
+        self.assertFalse(os.path.islink(os.path.join(scheduler_directory, identifier + '.lock')))
         shutil.rmtree(scheduler_directory)
 
+    def test_thundering_herd(self):
+        lockno = int(os.getenv('TTH_LOCKNO', str(sys.getrecursionlimit() + 16)))
+        scheduler_directory = tempfile.mkdtemp()
+        counter = os.path.join(scheduler_directory, 'counter')
+        class DummyST(ScheduledTask):
+            @defer.inlineCallbacks
+            def task(subself):
+                self.assertTrue(os.path.islink(os.path.join(scheduler_directory, subself.identifier + '.lock')))
+                with open(counter, 'w+') as fd:
+                    data = fd.read()
+                    fd.seek(0)
+                    if not data:
+                        fd.write('1')
+                        yield task.deferLater(reactor, 1, lambda: 42)
+                    else:
+                        # should not write anything as followup tasks should not be run
+                        fd.write(str(int(data) + 1))
+                        yield 34
+        identifier = "dummy"
+        dummy_st = DummyST(schedule='@daily', identifier=identifier, scheduler_directory=scheduler_directory)
+        dl = defer.DeferredList([dummy_st.run() for i in xrange(lockno)], consumeErrors=True)
+        @dl.addBoth
+        def so_what(results):
+            self.assertEqual(results[0], (True, None)) # wtf is `None` ?
```

Comment:
It looks like deadlock :) It's
```
[Failure instance: Traceback (failure with no frames): <class 'twisted.internet.defer.TimeoutError'>: <ooni.tests.test_scheduler.TestScheduler testMethod=test_thundering_herd> (test_thundering_herd) still running at 120.0 secs
]
```

Try running the testcase with `env TTH_LOCKNO=10` and `env TTH_LOCKNO=1000` -- different number of waiters produces different results. IMHO, that's exactly the reason of the deadlock I've seen at least couple of times — the lockfile had `mtime` equal to the moment when the long-running measurement was done and bunch of ~100 accumulated tasks waiting in a queue was started.